### PR TITLE
Add HTTP trace log publishing functionality

### PR DESCRIPTION
**Pull request by VijithaEkanayake**
Introduced trace log publishing functionality with json format.

## Purpose
Provide a way to publish HTTP trace logs to a socket in JSON format.

## Approach
Provide command line arguments to initialize HTTP trace log configurations and implement Json log formatter format Log records to json format.

## Automation tests
 - Unit tests - Yes

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
```
Command Line Arguments

-B[tracelog.http].level=TRACE -B[tracelog.http].publish.host="localhost" -B[tracelog.http].publish.port=5000 -B[tracelog.http].logto=socket

```

### CODE REVIEW

File: bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/BLogManager.java [hunk 1]
```diff
@@ -113,21 +119,53 @@ public BLogLevel getPackageLogLevel(String pkg) {
...
-    public void setHttpTraceLogHandler() {
-        Handler handler = new ConsoleHandler();
-        handler.setFormatter(new HTTPTraceLogFormatter());
-        handler.setLevel(Level.FINEST);
-
+    /**
+     * Sets the Http trace log handler.
+     *
+     * @throws IOException if there is I/O exception while creating a socket.
+     */
```
Comment (chamil321) [inline bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/BLogManager.java]: Document TraceLogConfigurationException too

File: bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/formatters/JsonLogFormatter.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                strbuff.append(
+                        "{ \"sourceClass\": \"" + frames[i].getClassName() + "\", \"sourceMethod\": \"" + frames[i]
+                                .getMethodName() + "\"," + " \"line\": \"" + frames[i].getLineNumber() + "\" }");
+                if (i < frames.length - 1) {
+                    strbuff.append(",");
+                }
+            }
+            strbuff.append("]");
+            strbuff.append("}");
+
```
Comment (chamil321) [inline bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/formatters/JsonLogFormatter.java]: Shall we remove this extra line?

File: bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/exceptions/TraceLogConfigurationException.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+*  KIND, either express or implied.  See the License for the
+*  specific language governing permissions and limitations
+*  under the License.
+*/
+package org.ballerinalang.logging.exceptions;
+
+/**
+ * This exception is thrown when an invalid configuration passed in to the Http trace logger.
+ *
+ * @since 0.965.0
```
Comment (pubudu91) [inline bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/exceptions/TraceLogConfigurationException.java]: 0.966.0

File: bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/formatters/JsonLogFormatter.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.Date;
+import java.util.TimeZone;
+import java.util.logging.Formatter;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+
+/**
+ * JSON log formatter for formatting HTTP trace log files.
+ *
+ * @since 0.965.0
```
Comment (pubudu91) [inline bvm/ballerina-logging/src/main/java/org/ballerinalang/logging/formatters/JsonLogFormatter.java]: 0.966.0

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpConnectionManager.java [hunk 5]
```diff
@@ -66,7 +68,13 @@ private HttpConnectionManager() {
                 .getServerBootstrapConfiguration(trpConfig.getTransportProperties());
 
         if (isHTTPTraceLoggerEnabled()) {
-            ((BLogManager) BLogManager.getLogManager()).setHttpTraceLogHandler();
+            try {
+                ((BLogManager) BLogManager.getLogManager()).setHttpTraceLogHandler();
+            } catch (IOException e) {
+                throw new BallerinaConnectorException("Invalid Http trace log parameters found.", e);
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpConnectionManager.java]: `HTTP trace log`*

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpConnectionManager.java [hunk 6]
```diff
@@ -66,7 +68,13 @@ private HttpConnectionManager() {
...
+            } catch (TraceLogConfigurationException e) {
+                throw new BallerinaConnectorException("Unsupported trace log configuration. " + e.getMessage(), e);
```
Comment (pubudu91) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpConnectionManager.java]: Shall we be specific here as well and say `HTTP trace log`?

**Comment by shafreenAnfar** [comment]
Talked to the Author and decided to close the PR. Author will send a new one.

**Pull request closed without merging**